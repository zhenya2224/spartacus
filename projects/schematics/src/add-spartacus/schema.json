{
  "$schema": "http://json-schema.org/schema",
  "$id": "SpartacusSchematics",
  "title": "Spartacus Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "items": {
        "enum": [
          "ASM",
          "Import-Export",
          "Saved-Cart",
          "Quick-Order",
          "CDC",
          "CDS",
          "Checkout",
          "Order",
          "Digital-Payments",
          "Administration",
          "Order-Approval",
          "Bulk-Pricing",
          "Image-Zoom",
          "Product-Variants",
          "VC-Configurator",
          "Textfield-Configurator",
          "CPQ-Configurator",
          "Qualtrics",
          "SmartEdit",
          "Store-Finder",
          "Personalization",
          "TMS-GTM",
          "TMS-AEPL",
          "Account",
          "Profile"
        ],
        "type": "string"
      },
      "default": [
        "ASM",
        "Checkout",
        "Order",
        "Import-Export",
        "Saved-Cart",
        "Quick-Order",
        "Image-Zoom",
        "Product-Variants",
        "SmartEdit",
        "Store-Finder",
        "Personalization",
        "Account",
        "Profile"
      ],
      "x-prompt": {
        "message": "Which Spartacus features would you like to set up?\nPlease note that for most Spartacus features to be properly configured, the Account feature is required.",
        "type": "list",
        "items": [
          {
            "value": "ASM",
            "label": "Assisted Services Module"
          },
          {
            "value": "Import-Export",
            "label": "Import/Export"
          },
          {
            "value": "Saved-Cart",
            "label": "Saved Cart"
          },
          {
            "value": "Quick-Order",
            "label": "Quick Order"
          },
          {
            "value": "CDC",
            "label": "Customer Data Cloud Integration"
          },
          {
            "value": "CDS",
            "label": "Context-Driven Services Integration"
          },
          {
            "value": "Checkout",
            "label": "Checkout"
          },
          {
            "value": "Order",
            "label": "Order"
          },
          {
            "value": "Digital-Payments",
            "label": "Digital Payments Integration"
          },
          {
            "value": "Administration",
            "label": "Organization - Adminstration (b2b feature)"
          },
          {
            "value": "Order-Approval",
            "label": "Organization - Order Approval (b2b feature)"
          },
          {
            "value": "Bulk-Pricing",
            "label": "Product - Bulk Pricing (b2b feature)"
          },
          {
            "value": "Product-Variants",
            "label": "Product - Variants"
          },
          {
            "value": "Image-Zoom",
            "label": "Product - Image Zoom"
          },
          {
            "value": "VC-Configurator",
            "label": "Product Configurator - Variant Configurator"
          },
          {
            "value": "Textfield-Configurator",
            "label": "Product Configurator - Textfield Configurator"
          },
          {
            "value": "CPQ-Configurator",
            "label": "Product Configurator - CPQ Configurator (b2b feature)"
          },
          {
            "value": "Qualtrics",
            "label": "Qualtrics"
          },
          {
            "value": "SmartEdit",
            "label": "SmartEdit"
          },
          {
            "value": "Store-Finder",
            "label": "Store Finder"
          },
          {
            "value": "Personalization",
            "label": "Tracking - Personalization"
          },
          {
            "value": "TMS-GTM",
            "label": "Tracking - Tag Management System - Google Tag Manager"
          },
          {
            "value": "TMS-AEPL",
            "label": "Tracking - Tag Management System - Adobe Experience Platform Launch"
          },
          {
            "value": "Account",
            "label": "User - Account"
          },
          {
            "value": "Profile",
            "label": "User - Profile"
          }
        ]
      }
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the Spartacus features.",
      "default": true
    },
    "baseUrl": {
      "type": "string",
      "description": "The url of the OCC backend",
      "default": "https://localhost:9002"
    },
    "occPrefix": {
      "type": "string",
      "description": "The OCC API prefix. E.g.: /occ/v2/"
    },
    "baseSite": {
      "type": "string",
      "description": "A comma-separated list of base site(s) to use with Spartacus"
    },
    "currency": {
      "type": "string",
      "description": "A comma-separated list of currencies to use in Spartacus",
      "default": "USD"
    },
    "language": {
      "type": "string",
      "description": "A comma-separated list of languages to use in Spartacus",
      "default": "en"
    },
    "urlParameters": {
      "type": "string",
      "description": "A comma-separated list that represents the order of site-context parameters. E.g.: ['baseSite', 'language', 'currency']"
    },
    "useMetaTags": {
      "type": "boolean",
      "description": "Configure baseUrl via meta tag in index.html"
    },
    "featureLevel": {
      "type": "string",
      "description": "Configure feature level for application."
    },
    "overwriteAppComponent": {
      "type": "boolean",
      "description": "Overwrite the contents of app.component.html file with the <cx-storefront> tag.",
      "default": true
    },
    "pwa": {
      "type": "boolean",
      "description": "Add PWA support to the project.",
      "default": false
    },
    "ssr": {
      "type": "boolean",
      "description": "Add Server-side rendering support to the project.",
      "default": false
    },
    "theme": {
      "type": "string",
      "description": "Select a style theme to add. E.g: santorini"
    }
  },
  "required": []
}
