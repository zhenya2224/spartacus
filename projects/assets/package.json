{
  "name": "@spartacus/assets",
  "version": "4.1.0-next.0",
  "homepage": "https://github.com/SAP/spartacus",
  "repository": "https://github.com/SAP/spartacus/tree/develop/projects/assets",
  "scripts": {
    "build": "ng build assets --configuration production && yarn generate:translations:ts-2-json",
    "generate:translations:properties-2-ts": "ts-node ./generate-translations-properties-2-ts && cd ../.. && npx prettier \"./projects/assets/src/translations/**/*.ts\" --write",
    "generate:translations:ts-2-json": "ts-node ./generate-translations-ts-2-json",
    "generate:translations:ts-2-properties": "ts-node ./generate-translations-ts-2-properties"
  },
  "dependencies": {
    "tslib": "^2.3.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
