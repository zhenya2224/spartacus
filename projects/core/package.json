{
  "name": "@spartacus/core",
  "version": "4.1.0-next.0",
  "description": "Spartacus - the core framework",
  "keywords": [
    "spartacus",
    "storefront",
    "angular",
    "framework"
  ],
  "homepage": "https://github.com/SAP/spartacus",
  "repository": "https://github.com/SAP/spartacus/tree/develop/projects/core",
  "license": "Apache-2.0",
  "dependencies": {
    "tslib": "^2.3.0"
  },
  "peerDependencies": {
    "@angular/common": "^12.0.5",
    "@angular/core": "^12.0.5",
    "@angular/platform-browser": "^12.0.5",
    "@angular/router": "^12.0.5",
    "@ngrx/effects": "^12.1.0",
    "@ngrx/router-store": "^12.1.0",
    "@ngrx/store": "^12.1.0",
    "angular-oauth2-oidc": "^10.0.1",
    "i18next": "^20.2.2",
    "i18next-http-backend": "^1.2.2",
    "rxjs": "^6.6.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
