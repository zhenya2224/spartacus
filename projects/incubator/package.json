{
  "name": "@spartacus/incubator",
  "version": "0.0.1",
  "description": "Spartacus incubator library",
  "dependencies": {
    "tslib": "^2.3.0"
  },
  "peerDependencies": {
    "@angular/core": "^12.0.5"
  }
}
