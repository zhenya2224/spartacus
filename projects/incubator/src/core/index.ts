export * from './incubator-core.module';
