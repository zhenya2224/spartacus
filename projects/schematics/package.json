{
  "name": "@spartacus/schematics",
  "version": "4.1.0-next.0",
  "description": "Spartacus schematics",
  "keywords": [
    "spartacus",
    "schematics"
  ],
  "homepage": "https://github.com/SAP/spartacus",
  "repository": "https://github.com/SAP/spartacus/tree/develop/projects/schematics",
  "license": "MIT",
  "main": "index.js",
  "scripts": {
    "build": "yarn clean && yarn --cwd ../../ generate:deps && ../../node_modules/.bin/tsc -p ./tsconfig.schematics.json",
    "clean": "../../node_modules/.bin/rimraf \"index.js*\" \"index.d.ts\" \"src/**/*.js\" \"src/**/*.js.map\" \"src/**/*.d.ts\"",
    "test": "yarn clean && ../../node_modules/.bin/jest --config ./jest.schematics.config.js"
  },
  "dependencies": {
    "@angular/pwa": "^12.0.5",
    "@nguniversal/express-engine": "^12.0.2",
    "semver": "^6.3.0",
    "ts-morph": "^9.1.0",
    "tslib": "^2.3.0"
  },
  "peerDependencies": {
    "@angular-devkit/core": "^12.0.5",
    "@angular-devkit/schematics": "^12.0.5",
    "@angular/compiler": "^12.0.5",
    "@schematics/angular": "^12.0.5",
    "jsonc-parser": "^3.0.0",
    "parse5": "^6.0.1",
    "typescript": "^4.2.4"
  },
  "ng-add": {
    "save": "devDependencies"
  },
  "ng-update": {
    "migrations": "./src/migrations/migrations.json",
    "packageGroup": [
      "@spartacus/core",
      "@spartacus/storefront",
      "@spartacus/styles",
      "@spartacus/user",
      "@spartacus/cds",
      "@spartacus/assets",
      "@spartacus/incubator",
      "@spartacus/organization",
      "@spartacus/product",
      "@spartacus/storefinder",
      "@spartacus/checkout",
      "@spartacus/asm",
      "@spartacus/smartedit",
      "@spartacus/qualtrics",
      "@spartacus/cdc",
      "@spartacus/tracking",
      "@spartacus/cart",
      "@spartacus/order",
      "@spartacus/digital-payments",
      "@spartacus/setup"
    ]
  },
  "schematics": "./src/collection.json"
}
