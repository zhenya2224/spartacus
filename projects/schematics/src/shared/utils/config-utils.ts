import { Node, SourceFile, ts as tsMorph } from 'ts-morph';
import {
  PROVIDE_CONFIG_FUNCTION,
  SPARTACUS_CORE,
  SPARTACUS_SETUP,
} from '../constants';
import { isImportedFromSpartacusLibs } from './import-utils';
import { CustomConfig } from './lib-utils';
import { getModule } from './new-module-utils';

export function getSpartacusProviders(sourceFile: SourceFile): Node[] {
  const module = getModule(sourceFile);
  if (!module) {
    return [];
  }
  const literal = module.getFirstDescendantByKind(
    tsMorph.SyntaxKind.ObjectLiteralExpression
  );
  const providers: Node[] = [];

  const properties =
    literal?.getChildrenOfKind(tsMorph.SyntaxKind.PropertyAssignment) ?? [];
  properties.forEach((property) => {
    if (
      property.getNameNode().getText() === 'providers' &&
      property.getInitializerIfKind(tsMorph.SyntaxKind.ArrayLiteralExpression)
    ) {
      const initializer = property.getInitializerIfKind(
        tsMorph.SyntaxKind.ArrayLiteralExpression
      );
      initializer?.getElements().forEach((element) => {
        if (Node.isCallExpression(element) || Node.isSpreadElement(element)) {
          const expression = element.getExpression();
          if (
            Node.isIdentifier(expression) &&
            isImportedFromSpartacusLibs(expression)
          ) {
            providers.push(element);
          }
        }
      });
    }
  });

  return providers;
}

const EMPTY_SPACE_REG_EXP = /\s+/gm;
export function normalizeConfiguration(config: string | Node): string {
  let newConfig = typeof config === 'string' ? config : config.getText();

  newConfig = newConfig.trim();

  if (newConfig.startsWith(PROVIDE_CONFIG_FUNCTION)) {
    newConfig = newConfig.replace(`${PROVIDE_CONFIG_FUNCTION}(`, '');
    newConfig = newConfig.substring(0, newConfig.length - 1);
  }

  newConfig = newConfig.replace(EMPTY_SPACE_REG_EXP, '');

  return newConfig;
}

export function getB2bConfiguration(): CustomConfig[] {
  return [
    {
      import: [
        {
          moduleSpecifier: SPARTACUS_CORE,
          namedImports: [PROVIDE_CONFIG_FUNCTION],
        },
        {
          moduleSpecifier: SPARTACUS_SETUP,
          namedImports: ['defaultB2bOccConfig'],
        },
      ],
      content: `provideConfig(defaultB2bOccConfig)`,
    },
    {
      import: [
        {
          moduleSpecifier: SPARTACUS_CORE,
          namedImports: [PROVIDE_CONFIG_FUNCTION],
        },
        {
          moduleSpecifier: SPARTACUS_SETUP,
          namedImports: ['defaultB2bCheckoutConfig'],
        },
      ],
      content: `provideConfig(defaultB2bCheckoutConfig)`,
    },
  ];
}
