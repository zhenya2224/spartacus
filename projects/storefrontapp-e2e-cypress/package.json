{
  "name": "storefrontapp-e2e-cypress",
  "version": "3.0.0",
  "main": "index.js",
  "author": "SAP, Spartacus team",
  "license": "Apache-2.0",
  "scripts": {
    "cy:open": "cypress open",
    "cy:open:b2b": "cypress open --env BASE_SITE=powertools-spa,OCC_PREFIX_USER_ENDPOINT=orgUsers",
    "cy:run": "cypress run",
    "cy:run:b2b": "cypress run --env BASE_SITE=powertools-spa,OCC_PREFIX_USER_ENDPOINT=orgUsers",
    "cy:run:core": "cypress run --spec \"cypress/integration/**/*.core-e2e-spec.ts\"",
    "cy:run:ci": "cypress run --config-file $CYPRESS_CONFIG --record --key $CYPRESS_KEY --tag \"b2c,all\" --parallel --group B2C --spec \"cypress/integration/!(vendor|b2b|ssr)/**/*e2e-spec.ts\"",
    "cy:run:ci:core": "cypress run --config-file $CYPRESS_CONFIG --record --key $CYPRESS_KEY --tag \"b2c,core\" --parallel --group B2C --spec \"cypress/integration/!(vendor|b2b|ssr)/**/*.core-e2e-spec.ts\"",
    "cy:run:ci:core:b2b": "cypress run --config-file $CYPRESS_CONFIG --record --key $CYPRESS_KEY --tag \"b2b,core\" --parallel --group B2B --spec \"cypress/integration/b2b/regression/**/*.core-e2e-spec.ts\"",
    "cy:run:ci:ssr": "cypress run --config-file $CYPRESS_CONFIG --config baseUrl=http://localhost:4000 --record --key $CYPRESS_KEY --tag \"ssr,all\" --parallel --group SSR --spec \"cypress/integration/ssr/pages.core-e2e-spec.ts\"",
    "cy:run:ci:b2b": "cypress run --config-file $CYPRESS_CONFIG --record --key $CYPRESS_KEY --tag \"b2b,all\" --parallel --group B2B --spec \"cypress/integration/b2b/regression/**/*e2e-spec.ts\"",
    "cy:run:ci:cds": "cypress run --config-file $CYPRESS_CONFIG --record --key $CYPRESS_KEY --tag \"2011,b2c,all-cds\" --group CDS --spec \"cypress/integration/vendor/cds/**/*.core-e2e-spec.ts\"",
    "cy:run:product-configurator-vc": "cypress run --spec \"cypress/integration/**/*-configurator*.*e2e-spec.ts\"",
    "cy:run:cpq": "cypress run  --config-file cypress.cpq.json --spec \"cypress/integration/**/cpq-configuration.core-e2e-spec.ts\"",
    "cy:run:ci:cdc": "cypress run --config-file $CYPRESS_CONFIG --record --key $CYPRESS_KEY --tag \"2005,cdc\" --group CDC --spec \"cypress/integration/vendor/cdc/**/*.e2e-spec.ts\"",
    "cy:run:ci:digital-payments": "cypress run --config-file $CYPRESS_CONFIG --record --key $CYPRESS_KEY --tag \"2105,digital-payments\" --group DIGITAL-PAYMENTS --spec \"cypress/integration/vendor/digital-payments/*.e2e-spec.ts\""
  },
  "devDependencies": {
    "@cypress/webpack-preprocessor": "^5.9.1",
    "@testing-library/cypress": "^8.0.1",
    "cypress": "^8.5.0",
    "cypress-file-upload": "^5.0.8"
  }
}
