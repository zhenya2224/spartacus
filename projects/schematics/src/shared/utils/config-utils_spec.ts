/// <reference types="jest" />

import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import {
  Schema as ApplicationOptions,
  Style,
} from '@schematics/angular/application/schema';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import * as path from 'path';
import { InMemoryFileSystemHost, Project, SourceFile } from 'ts-morph';
import { getSpartacusProviders } from './config-utils';

const collectionPath = path.join(__dirname, '../../collection.json');
const schematicRunner = new SchematicTestRunner('schematics', collectionPath);

describe('Storefront config utils', () => {
  let appTree: UnitTestTree;
  const workspaceOptions: WorkspaceOptions = {
    name: 'workspace',
    version: '0.5.0',
  };
  const appOptions: ApplicationOptions = {
    name: 'schematics-test',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: Style.Scss,
    skipTests: false,
    projectRoot: '',
  };
  const defaultOptions = {
    project: 'schematics-test',
  };

  beforeEach(async () => {
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      )
      .toPromise();
    appTree = await schematicRunner
      .runSchematicAsync('add-spartacus', defaultOptions, appTree)
      .toPromise();
  });

  describe('ts-morph config utils', () => {
    const configFileContent = `
import { NgModule } from '@angular/core';
import {
CartAddEntrySuccessEvent,
CartRemoveEntrySuccessEvent,
provideConfig,
} from '@spartacus/core';
import { NavigationEvent, defaultCmsContentProviders } from '@spartacus/storefront';
import { PersonalizationRootModule } from '@spartacus/tracking/personalization/root';
import { AepModule } from '@spartacus/tracking/tms/aep';
import { BaseTmsModule, TmsConfig } from '@spartacus/tracking/tms/core';
import { GtmModule } from '@spartacus/tracking/tms/gtm';
import { someFunction, someSpread } from '@some/package';

@NgModule({
imports: [
BaseTmsModule.forRoot(),
GtmModule,
AepModule,
PersonalizationRootModule,
],
providers: [
someFunction(),
...someSpread,
...defaultCmsContentProviders,
provideConfig(<TmsConfig>{
  tagManager: {
    gtm: {
      events: [NavigationEvent, CartAddEntrySuccessEvent],
    },
    aep: {
      events: [NavigationEvent, CartRemoveEntrySuccessEvent],
    },
  },
}),
provideConfig({
  featureModules: {
    personalization: {
      module: () =>
        import('@spartacus/tracking/personalization').then(
          (m) => m.PersonalizationModule
        ),
    },
  },
}),
],
})
export class TrackingFeatureModule {}
`;
    let sourceFile: SourceFile;

    beforeAll(() => {
      const project = new Project({
        fileSystem: new InMemoryFileSystemHost(),
      });
      sourceFile = project.createSourceFile('test.ts', configFileContent);
    });

    describe('getSpartacusProviders', () => {
      it('should return all providers from spartacus in file', () => {
        const providers = getSpartacusProviders(sourceFile);
        expect(providers.length).toEqual(3);
        expect(providers[0].getText()).toMatchSnapshot();
        expect(providers[1].getText()).toMatchSnapshot();
        expect(providers[2].getText()).toMatchSnapshot();
      });
    });
  });
});
