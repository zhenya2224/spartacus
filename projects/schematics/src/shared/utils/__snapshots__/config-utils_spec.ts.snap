// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storefront config utils ts-morph config utils getSpartacusProviders should return all providers from spartacus in file 1`] = `"...defaultCmsContentProviders"`;

exports[`Storefront config utils ts-morph config utils getSpartacusProviders should return all providers from spartacus in file 2`] = `
"provideConfig(<TmsConfig>{
  tagManager: {
    gtm: {
      events: [NavigationEvent, CartAddEntrySuccessEvent],
    },
    aep: {
      events: [NavigationEvent, CartRemoveEntrySuccessEvent],
    },
  },
})"
`;

exports[`Storefront config utils ts-morph config utils getSpartacusProviders should return all providers from spartacus in file 3`] = `
"provideConfig({
  featureModules: {
    personalization: {
      module: () =>
        import('@spartacus/tracking/personalization').then(
          (m) => m.PersonalizationModule
        ),
    },
  },
})"
`;
